packages/firestore/src/api/credentials.ts

```diff
@@ -188,44 +215,62 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
   private tokenCounter = 0;
 
   /** The listener registered with setChangeListener(). */
-  private changeListener: CredentialChangeListener | null = null;
+  private changeListener: CredentialChangeListener = () => Promise.resolve();
 
   private forceRefresh = false;
 
-  private auth: FirebaseAuthInternal | null;
+  private auth: FirebaseAuthInternal | null = null;
+
+  private asyncQueue: AsyncQueue | null = null;
 
   constructor(authProvider: Provider<FirebaseAuthInternalName>) {
     this.tokenListener = () => {
       this.tokenCounter++;
+      this.receivedUser.resolve();
       this.currentUser = this.getUser();
-      this.receivedInitialUser = true;
-      if (this.changeListener) {
-        this.changeListener(this.currentUser);
+      if (this.initialEventRaised && this.asyncQueue) {
+        // We only invoke the change listener here if the initial event has been
+        // raised. The initial event itself is invoked synchronously in
+        // `awaitTokenAndRaiseInitialEvent()`.
+        this.asyncQueue.enqueueRetryable(() => {
+          this.changeListener(this.currentUser);
```

Comment:
In the `registerAuth()`, we do 
```
        this.awaitTokenAndRaiseInitialEvent();
        this.auth.addAuthTokenListener(this.tokenListener);
```

We already invoke `this.changeListener` in the asyncQueue in `awaitTokenAndRaiseInitialEvent`, should we not do it again in `this.tokenListener`?